README.md
# Wecab 

### Web Content Aggregation Bot 网络内容聚合机器人  
#### 最好的综合型订阅制机器人  
前身是给[CQ-picfinder-robot](https://github.com/Tsuk1ko/CQ-picfinder-robot)做的小挂件（已完全脱离）  
如果有bug请复制pm2log中的错误信息并提交issus，我会尽快解决（可能） 
觉得好用不妨点个Star谢谢  

[Wiki](https://github.com/Ninzore/Wecab/wiki)  
[更新日志](https://github.com/Ninzore/Wecab/blob/master/CHANGELOG.md)

## 特色  
1. 所有不需要申请任何api，包括微博，B站，Twitter，翻译，即插即用  
2. 群组分离，互不干扰  

**注意！！！**  
**近期会支持嵌入式数据库**如果不想用MongoDB的人可以等一阵子(可能你们得等久点了...)  
  
目前需要安装MongoDB，没装的参考官方教程https://docs.mongodb.com/manual/installation/ ，端口是默认端口不要改，然后使用mongorestore将dump文件夹中的东西恢复到mongodb中，这样才能使用小游戏功能，当然如果不用的话也可以，删除main中所有pokemon相关的语句就行  

## 部件列表  
正在缓慢更新中  
TODO:  
1. Pixiv订阅  
2. 接入RSSHub 
3. 从MongoDB换到SQLite或者LevelDB

### 1. 看微博，B站，Twitter动态  
用 “看看**谁谁上条**微博/B站/Twitter/推特”，谁谁是要看的人的名字，上条可以改，可以要求看置顶，上条，上上条，上上上条.....，也可以直接写第x条，x范围是0~9  
反微博小程序，会自动把微博小程序内容给抓出来，造福PC用户  
也可以用 **看看 url**，url为单条微博/B站/Twitter的网址，在app里，选分享链接就好，支持b23短链接  

### 2. 订阅微博，B博，Twitter
每当有新内容时会转发到群内  
  
### 3. 教说话  
教bot对任意关键词做出反应 精确/模糊/正则  
  
### 4. 教复读  
教bot复读任意语句 精确/模糊/正则  

### 5. Pixiv看图  
***看看p站***后接p站图片id或者包含id的url，如果该id有多图会自动合并发送，最多同时发送9张  

### 6. 翻译  
**翻译>** 后接需要翻译的外语句子，自动识别语言并翻译成中文  
**中译x>** x为目标语言，可以是日韩英法德俄，后接中文，会翻译为目标语言  
目前使用腾讯翻译

### 7. 骰子  
可以扔出任意面数（<1000），任意个数(<10)的骰子，可以指定最小值，自动统计总分  
用法的话，参考  
.dice6  扔一个D6  
.dice8x5  扔5个D8  
.dice16,10  扔1个最低为10的D16  
.dice20x3,10  扔3个最低为10的D20  

### 8.简单直接宝可梦  
**注意！！！** 使用前必须先将下载dump然后用mongorestore存进mongoDB  
一个极度简化版的群内对战游戏


CHANGELOG.md
# 更新日志
  
## 2021

### 09/09 v1.8.27
* plugin/pixivImage:
  * Fix: 多图时无法正常发送
  * Remove: 图片大小限制

### 08/29 v1.8.26
* plugin/pixivImage:
  * Fix: 多图时无法正常发送

### 07/17 v1.8.25
* plugin/translate:
  * Fix: 修复reauth失效

### 06/22 v1.8.24
* plugin/zen:
  * Fix: 禁言无效

### 05/10 v1.8.23
* plugin/weibo:
  * Fix: 链接格式为`m.weibo.cn/detail/\d`时id获取出错

### 05/02 v1.8.22
* plugin/twitter:
  * Fix: 视频和图片发送失败（CQ码错误）

### 04/24 v1.8.2
* plugin/twitter:
  * Feat: 支持使用代理下载多媒体内容
  * Fix: for let
* utils/download:
  * Fix: 路径错误，axios代理错误
* utils/initialise:
  * Feat: 初始化更多工具

### 04/24 v1.8.0
* plugin/bilibili:
  * Fix: 订阅数量超过一定时频繁412error
* plugin/twitter:
  * Feat: 移动代理设置到axiosProxied

### 04/16 v1.7.9
* main & utils/userManagement:
  * Feat: 增加取消管理员指令 `desu`

### 04/15 v1.7.8
* utils/axiosProxied.js:
  * Feat: axios代理
* utils/download.js
  * Feat: 更多可下载文件格式
  * Feat: 不支持的格式会提示，失败时会显示链接
  * Style: esm

### 04/13 v1.7.7
* main:
  * Feat: 可以使用指令增加管理员了
  * Change: 原Logger仅作为复读计数使用，其他功能移动到userManagement

### 04/12 v1.7.6
* utils/userManagement:
  * Feat: 增加SU用户列表（在data/userControl.json）
* main:
  * Feat: 把SU列表里的人权限提高
  * Style: 缩进

### 04/11 v1.7.53
* Chore: move around
  * move modules => utils
  * move modules/plugin => plugin

### 04/10 v1.7.52
* Feat: 更新指令
  * pm2start => start
  * pm2stop => stop
  * pm2log => log
  * start => use node .
  * test => (remove)

### 04/09 v1.7.51
* modules/plugin/weibo:
  * Fix: 视频链接未转义导致CQ码错误视频发送失败

### 04/04 v1.7.5
* modules/plugin/weibo:
  * Feat: 新增权限检查
  * Feat: 新增停启判断
  * Style: add semicolon
* modules/config:
  * Rename: configStorage => config
* utils/initialise
  * Feat: permissionCheck
* main:
  * Feat: 可以设定是否检查时间线
* config.default.json:
  * 新增 weibo, bilibili, twitter, mongoDB  
    learn字段移出bot,  
    bot字段新增noPermission

### 04/01 v1.7.41
* Main:
  * Feat: 提高admin的权限

### 03/31 v1.7.4
* Main:
  * Fix: ban对提醒消息未生效

### 03/30 v1.7.31
* Twitter:
  * Fix: 我是SB
  * Change: 规范报错log格式

### 03/29 v1.7.3
* Twitter:
  * Fix: 代理未正确运行
* config:
  * Change: 修改proxy字段（Object -> String)

### 03/26 v1.7.2
* Bilibili:
  * Fix: 获取动态列表失败时读取了undefined
  * Chore: 增加名字搜索失败时的报错信息

### 03/22 v1.7.1
* Translate:
  * Feat: 自动获取reauth链接
  * Fix: typo

### 03/17 v1.7.02
* Twitter:
  * Fix: 没有正确判断tweet类型（回复判断为原创）

### 03/17 v1.7.0
* Twitter:
  * Refactor: 检查更新机制更新
  * Remove: 多余代码块
  * Remove: 不再检查图片大小
  * Fix: 转发时出现过多条Tweet
  * Style: 变量重命名

### 03/16 v1.6.8
* Twitter:
  * Fix: Twitter获取时间线失效
  * Style: 加上async标识

### 03/11 v1.6.7
* Translate:
  * Fix: reauth地址更新
  * Reformat: 修改逻辑，在失败后不会一直尝试重新初始化

### 03/01 v1.6.6
* Twitter:
  * Feat: 视频内容改为用小视频形式发送 (需要后端支持，如go-cqhttp v0.9.38以上)
  * Refactor: 刷新周期改为固定30秒一次
  * Fix: 微博没有订阅时仍然尝试读取
  * Style: 增加async标识

### 02/28 v1.6.5
* Twitter:
  * Feat: 视频内容改为用小视频形式发送 (需要后端支持，如go-cqhttp v0.9.38以上)
  * Feat: 排列Twitter内容时增加错误处理，失败时报错
  * Doc: 修改format的Doc

### 02/09 v1.6.4
* Twitter:
  * Fix: Twitter card的summary部分可能没有description导致的无法转发

### 02/05 v1.6.3
* telephone:
  * Feat: 添加公告（给黄页中所有人发消息）

### 01/30 v1.6.2
* twitter:
  * Fix: 搜索时由于cookie刷新周期过长，token失效导致的失败

### 01/26 v1.6.1
* telephone:
  * Feat: 添加黄页相关功能（拨号，按名字发送）
  * Fix: 发送消息时本群必须处于开放状态

### 01/26 v1.6.0
* dice:
  * Remove: 多余代码
  * Fix: 未指定最小值时骰子可能为0

  ### 01/25 v1.5.9
* translate:
  * Fix: CPU带宽过量占用
  * Fix: CQ码和链接之后会被无视

### 01/24 v1.5.8
* pretendLearn:
  * fix: 当模式为正则时未正确处理转义导致使用[]时正则失效

### 01/22 v1.5.7
* telephone:
  * feat: 可以自行禁止接收消息，使用`开启/关闭跨组消息`控制

### 01/21 v1.5.6
* feat: 使用“发送<组号>，blabla”，可以跨组发消息了

### 01/10 v1.5.5
* pretendLearn: 
  * feat: 对于复读和教学，**没有**指定模式时，如果满足以下条件会变为精确匹配: 
    1. 如果句子长度大于20
    2. 包含CQ码

### 01/09 v1.5.4
* main: 
  * feat: 增加一些全局功能
*  initilise:
  * feat: axios增加重试机制

### 01/07 v1.5.3
* pretendLearn: 
  * Fix: 错误的正则并没有正确地报错跳出
  * Feat: 教学时如果没有指定模式，rhs没有cq码，并且字数大于20，会自动变为精确触发

### 01/03 v1.5.2
* Bilibili: 
  * Fix: 增加延迟，降低412错误出现率

## 2020

### 12/08 v1.5.1
* Twitter: 
  * Fix: 发现检测连接的时候可能需要额外的请求头，要不然可能会失败

### 12/08 v1.5.0
* PretendLearn: 
  * Fix: 建立等式两边不再允许使用图片，总感觉有问题
  * Feat: 等式替换时会无视大小写
  * Feat: 教学时可以使用多行文字
  * Feat: 教学时如果没有指定模式，并且应答词长度大于20会默认使用精确

### 11/25 v1.4.9
* Translate: 自24日起，观测到该模块可能会引起CPU以及带宽占用缓慢上升直到100%，暂时移除，等待修复

### 11/20 v1.4.8
* Remove: 把广播功能删了

### 11/20 v1.4.7
* Twitter
  * fix: 好像之前proxy并未起效
* 注: package新增https-proxy-agent，注意安装

### 11/18 v1.4.6
* Translate
  * feat: 适配腾讯翻译新的鉴权机制
  * fix: 翻译文字未对特殊字符进行转义
  * chore: 暴露翻译接口以供别的引用使用

### 11/15 v1.4.5
* Twitter
  * Update: Twitter整体改版（未经过大量验证）
* Bilibili
  * Fix: 有时无法获取专栏封面

### 11/14 v1.4.0
* Bilibili
  * feat: 增加对直播间的解析

### 10/29 v1.3.9
* Twitter
  * feat: 增加proxy设置
* config file
  * feat: 增加proxy设置栏

### 10/25 v1.3.8
* pokemon
  * Fix: 一堆数据库调用错误
* 更新mongodb版本

### 10/24 v1.3.7
* nbnhhsh
  * Update: 加回来了，改了一下用法

### 10/18 v1.3.6
* pokemon
  * Fix: 
      1. 地点错误
      2. 对战结算错误
  * Update: 更改低保模式为自动

### 10/15
* weibo:
  * Update: 支持/status的网址


plugin/weibo.js
import axios from 'axios';
const mongodb = require('mongodb').MongoClient;

const CONFIG = global.config.weibo;
const db_path = global.config.mongoDB;
const PERMISSION = CONFIG.permission;

function unEscape(str) {
    const label = {"#44": ",", "#91": "[", "#93": "]", "amp": "&"}
    return str.replace(/&(#44|#91|#93|amp);/g, (_, s) => {
        return label[s];
    })
}

// function loadFilter() {
    
// }

/**
* @param {number} uid 用户uid
* @param {number} mid 单条微博mid
* @returns {object} http header
*/
function httpHeader(uid = 0, mid = 0) {
    let containerid = "107603" + uid;
    let since_id = mid;
    
    let headers = {
        "Host": "m.weibo.cn",
        "scheme": "https",
        "User-Agent": "Mozilla/5.0 (iPhone; CPU iPhone OS 12_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148",
        "Accept":'application/json, text/plain, */*',
        "X-Requested-With":"XMLHttpRequest",
    }

    let params = {
        "value": uid,
        "containerid": containerid,
    };
    //携带参数
    if (since_id != 0) params["since_id"] = since_id;

    let payload = {
        headers : headers,
        params : params
    }

    return payload;
}

/**
* @param {string} user_name 用户名
* @returns {Promise} number 用户uid，如果搜索不到返回false
*/
async function getUserId(user_name = "") {
    user_name = encodeURI(user_name);
    return axios({
        method:'GET',
        url: "https://m.weibo.cn/api/container/getIndex",
        headers : httpHeader().headers,
        params : {
            "containerid" : "100103type=3&q=" + user_name,
            "page_type" : "searchall"
        }
    }).then(response => {
        if (response.data.ok != 1) {
            return false;
        }
        else if (response.data.data.cards.length > 0) return response.data.data.cards[1].card_group[0].user.id;
    })
    .catch(err => {
        console.error("weibo getUserId error code = ", err.code);
        return false;
    });
}

/**
 * 从用户时间线获取微博
 * @param {number} uid 用户uid
 * @param {number} num 需要获取的微博，-1为查找最新，-2为查找置顶，0为置顶或者最新，1是次新，以此类推，只允许0到9
 * @returns {Promise} 单条微博mblog
*/
async function getTimeline(uid, num = -1) {
    let payload = httpHeader(uid);
    return axios({
            method:'GET',
            url: "https://m.weibo.cn/profile/info",
            params : {uid : uid},
            headers: payload.headers
        }).then(response => {
            if (num == -2) {
                if ("isTop" in response.data.data.statuses[0] && response.data.data.statuses[0].isTop == 1) {
                    return response.data.data.statuses[0];
                }
                else response.data.data.statuses[0];
            }
            if (num == -1) {
                let card_num_seq = [0,1,2,3,4,5,6,7,8,9];
                let temp = 0;
                for (let i=0; i<response.data.data.statuses.length - 1; i++) {
                    if (response.data.data.statuses[i].id < response.data.data.statuses[i+1].id) {
                        temp = card_num_seq[i]
                        card_num_seq[i] = card_num_seq[i+1];
                        card_num_seq[i+1] = temp;
                    }
                }
                return response.data.data.statuses[card_num_seq[0]];
            }
            else return response.data.data.statuses[num];
        }).catch(err => {
            console.error(err.code + "\nweibo getTimeline error, uid= " + uid);
            return false;
        });
}

/**
 * 增加订阅
 * @param {number} uid 微博用户id
 * @param {string} option 偏好设置
 * @param {object} context
 * @returns {boolean} 错误返回false
 */
function subscribe(uid, option, context) {
    let group_id = context.group_id;
    let option_nl = opt2optnl(option);
    mongodb(db_path, {useUnifiedTopology: true}).connect().then(async mongo => {
        let coll = mongo.db('bot').collection('weibo');
        let weibo = await coll.find({weibo_uid : uid}).toArray();
        if (weibo.length == 0) {
            let mblog = await getTimeline(uid);
            if (mblog == undefined) return false;
            let mid = mblog.mid;
            let screen_name = mblog.user.screen_name;
            coll.insertOne({weibo_uid : uid, name : screen_name, mid : mid, groups : [group_id], [group_id] : option},
                (err) => {
                    if (err) console.error(err + "\nweibo database subscribes insert error");
                    else replyFunc(context, `已订阅${screen_name}的微博，模式为${option_nl}`, true);
                    mongo.close();
                });
        }
        else {
            coll.findOneAndUpdate({weibo_uid : uid},
                                  {$addToSet : {groups : group_id}, $set : {[group_id] : option}},
                (err, result) => {
                    if (err) console.error(err + "\nweibo database subscribes update error");
                    else {
                        let text = "";
                        if (result.value.groups.includes(group_id)) text = "多次订阅有害我的身心健康";
                        else text = `已订阅${result.value.name}的微博，模式为${option_nl}`;
                        replyFunc(context, text, true);
                        mongo.close();
                    }
            });
        }
    }).catch(err => console.error(err + "weibo subscribe error, uid= " + uid));
}

/**
 * 取消订阅
 * @param {string} name 微博用户名
 * @param {object} context
 * @returns {} no return
 */
function unSubscribe(name, context) {
    let group_id = context.group_id;
    let name_reg = new RegExp(name, 'i')
    mongodb(db_path, {useUnifiedTopology: true}).connect().then(async mongo => {
        let coll = mongo.db('bot').collection('weibo');
        await coll.findOneAndUpdate({name : name_reg},
                                    {$pull : {groups : {$in : [group_id]}}, $unset : {[group_id] : []}},
            async (err, result) => {
                if (err) console.log(err + "database subscribes delete error");
                else {
                    let text = "";
                    if (result.value == null || !result.value.groups.includes(group_id)) text = "小火汁你压根就没订阅嗷";
                    else {
                        text = "已取消订阅" + result.value.name + "的微博";
                        if (result.value.groups.length <= 1) await coll.deleteOne({_id : result.value._id});
                    }
                    replyFunc(context, text, true);
                }
            mongo.close();
        });
    }).catch(err => console.error(err + "weibo unsubscribe error, uid= " + uid));
}

/**
 * 每过x分钟检查一次订阅列表，如果订阅一个微博账号的群的数量是0就删除
 */
function checkWeiboDynamic() {
    let check_interval = 6 * 60 * 1000;
    let i = 0;
    // const AUTO_PUSH_FILTER = CONFIG.autoPushfilter;
    setInterval(() => {
        mongodb(db_path, {useUnifiedTopology: true}).connect().then(async mongo => {
            let coll = mongo.db('bot').collection('weibo');
            let subscribes = await coll.find({}).toArray();
            mongo.close();
            i = 0;
            if (subscribes && subscribes.length > 0) {
                check_interval = subscribes.length * 1000 * 30;
                checkEach();
            }

            function checkEach() {
                setTimeout(async function() {
                    try {
                        let stored_info = subscribes[i];
                        let mblog = await getTimeline(stored_info.weibo_uid);
                        let last_mid = stored_info.mid;
                        let current_mid = mblog.mid;
    
                        if (current_mid > last_mid) {
                            let groups = stored_info.groups;
                            groups.forEach(group_id => {
                                if (checkOption(mblog, stored_info[group_id])) {
                                    const context = {group_id : group_id, message_type : "group"};
                                    format(mblog, context).then(payload => replyFunc(context, payload)).catch(err => console.error(err));
                                }
                                else;
                            });
                            mongodb(db_path, {useUnifiedTopology: true}).connect().then(async mongo => {
                                let coll = mongo.db('bot').collection('weibo');
                                coll.updateOne({weibo_uid : stored_info.weibo_uid},
                                    {$set : {mid : current_mid}}, 
                                    (err, result) => {
                                        if (err) console.error(err + " database update error during checkWeibo");
                                        mongo.close();
                                    });
                            }).catch(err => console.error(err));
                        }
                        i++;
                        if (i < subscribes.length) checkEach();
                    } catch(err) {
                        console.error(err, '\n', subscribes[i]);
                        i++;
                        if (i < subscribes.length) checkEach();
                    }
                }, (check_interval-subscribes.length*1000)/subscribes.length);
            }
        }).catch(err => console.error(err));
    }, check_interval);

    function checkOption(mblog, option) {
        if (option == "all") return true;
        let status = "";
        if ("retweeted_status" in mblog) status = "with_rt";
        else if (mblog.pic_num > 0 && "pics" in mblog) status = "ori_with_pic";
        else status = "origin"

        if (option == "rt_only" && status == "with_rt") return true;
        else if (option == "pic_only" && status == "ori_with_pic") return true;
        else if (option == "origin") {
            if (status == "ori_with_pic" || status == "origin") return true;
        }
        else return false;
    }
}

/**
 * @param {object} context
 * @returns {} no return
 */
function checkWeiboSubs(context) {
    let group_id = context.group_id;
    mongodb(db_path, {useUnifiedTopology: true}).connect().then(async mongo => {
        let coll = mongo.db('bot').collection('weibo');
        await coll.find({groups : {$elemMatch : {$eq : group_id}}}, {projection: {_id : 0}})
            .toArray().then(result => {
                if (result.length > 0) {
                    let name_list = [];
                    result.forEach(weibo_obj => {
                        let option_nl = "仅原创";
                        option_nl = opt2optnl(weibo_obj[group_id])
                        name_list.push(`${weibo_obj.name}，${option_nl}`);
                    });
                    let subs = "本群已订阅:\n" + name_list.join("\n");
                    replyFunc(context, subs, true);
                }
                else replyFunc(context, "你一无所有", true);
            })
        mongo.close();
    }).catch(err => console.error(err + " weibo checkWeiboSubs error, group_id= " + group_id));
}

/**
 * @param {object} context
 * @returns {} no return
 */
function clearSubs(context, group_id) {
    mongodb(db_path, {useUnifiedTopology: true}).connect().then(async mongo => {
        let coll = mongo.db('bot').collection('weibo');
        try {
            let matchs = await coll.find({groups : {$in : [group_id]}}).toArray();
            if (matchs.length < 1) {replyFunc(context, `未见任何微博订阅`); return;}
            for (let item of matchs) {
                let res = await coll.findOneAndUpdate({_id : item._id}, {$pull : {groups : {$in : [group_id]}}, $unset : {[group_id] : []}}, {returnOriginal : false});
                if (res.value.groups.length < 1) await coll.deleteOne({_id : res.value._id});
            }
            replyFunc(context, `清理了${matchs.length}个微博订阅`);
        }
        catch(err) {
            console.error(err);
            replyFunc(context, '中途错误，清理未完成');
        }
        finally {mongo.close();}
    }).catch(err => console.error(err + " weibo checkWeiboSubs error, group_id= " + group_id));
}

/**
 * @param {string} text 需要过滤的html
 * @returns {string} 处理完成
 */
function textFilter(text) {
    return text.replace(/[\r\n]/g, "")
                .replace(/<a href="\/status\/.*\d">/g, "")
                .replace(/<a href='\/n\/.+?'>(.+?)<\/a>/g, "$1")  //@
                .replace(/<a  href="https:\/\/m.weibo.cn\/search.+?<span class="surl-text">(.+?)<\/span><\/a>/g, "$1")  //tag
                .replace(/<a  href="https:\/\/m.weibo.cn\/p\/index\?extparam.+?<span class="surl-text">(.+?)<\/span><\/a>/g, "[$1超话]")  //超话
                .replace(/<span class="url-icon"><img alt=(\[.+?\]).+?\/><\/span>/g, "$1")  //表情
                .replace(/<a data-url=\\?\"(.*?)\\?\".*?<\/a>/g, "$1")
                .replace(/<a data-url=.*?href=\\?"(.*?)".*?>/g, '$1')
                .replace(/<img style=.*?>/g, "")
                .replace(/<span.+?span>/g, "")
                .replace(/<a.+<\/a>/g, "")
                .replace(/<br \/>/g , "\n")
                .replace(/&quot;/g , "'")
                .replace(/&gt;/g , ">")
                .replace(/&lt;/g , "<")
                .replace(/网页链接/g, "")
                .replace(/\\.*?秒拍视频/g, "");
}

/**
* @param {object} mblog 单条微博mblog，可以是Promise
* @param {boolean} textForm 是否整理为成string
* @returns {Promise} Promise，由文字，图片，链接集合的单条微博的array或string
*/
async function format(mblog, context) {
    mblog = await mblog;
    let payload = [`${mblog.user.screen_name}的微博`];
    let text = mblog.text;
    if (/<a.+>全文<\/\a>/.test(text)) text = await weiboText(mblog.id);
    payload.push(textFilter(text));
    if ("pics" in mblog) {
        let pics = mblog.pics;
        let pic_str = "";
        let pic_url = "";
        for (let pic of pics) {
            pic_url = pic.large.url;
            pic_str += `[CQ:image,cache=0,file=${pic_url}]`;
        }
        payload.push(pic_str);
    }
    if ("page_info" in mblog) {
        if("media_info" in mblog.page_info){
            let media = mblog.page_info.media_info;
            let media_src = "";
            if ("hevc_mp4_hd" in media && media.hevc_mp4_hd != "") media_src += media.hevc_mp4_hd;
            else if ("h265_mp4_hd" in media && media.h265_mp4_hd != "") media_src += media.h265_mp4_hd;
            else if ("mp4_720p_mp4" in media && media.mp4_720p_mp4 != "") media_src += media.mp4_720p_mp4;
            else if ("mp4_hd_url" in media && media.mp4_hd_url != "") media_src += media.mp4_hd_url;
            else media_src += media.stream_url;
            payload.push(`[CQ:image,cache=0,file=${mblog.page_info.page_pic.url}]`);
            if (context) replyFunc(context, `[CQ:video,file=${media_src.replace(",", "&#44;")}]`);
        }
    }

    if ("retweeted_status" in mblog) {
        let rt_weibo = await format(mblog.retweeted_status, false);
        payload = payload.concat("转发自: " + rt_weibo)
    }

    return payload.join("\n");
}

/**
 * @param {string} option 偏好设置
 * @returns {string} 中文偏好设置 
 */
function opt2optnl(option = "origin") {
    let option_nl = "仅原创"
    switch (option) {
        case "origin": option_nl = "仅原创"; break;
        case "rt_only": option_nl = "仅转发"; break;
        case "pic_only": option_nl = "只看图"; break;
        case "all": option_nl = "全部"; break;
        default: break;
    }
    return option_nl;
}

/**
 * @param {string} option_nl 中文偏好设置
 * @returns {string} 偏好设置 
 */
function optnl2opt(option_nl = "仅原创") {
    let option = "origin";
    switch (option_nl) {
        case "仅原创": option = "origin"; break;
        case "仅转发": option = "rt_only"; break;
        case "只看图": option = "pic_only"; break;
        case "全部": option = "all"; break;
        default: break;
    }
    return option;
}

/**
 * @param {string} id 微博id
 * @returns {Promise} 单条微博的完整文字部分
 */
function weiboText(id) {
    return axios.get("https://m.weibo.cn/statuses/extend?id=" + id, {headers : httpHeader().headers})
                .then(res => res.data.data.longTextContent)
                .catch(err => {return err.response.status})
}

/**
 * @param {string} id 微博id
 * @param {object} context
 * @returns {} no return
 */
function rtSingleWeibo(id, context) {
    axios.get("https://m.weibo.cn/statuses/show?id=" + id, {headers : httpHeader().headers})
    .then(async res => {
        let payload = await format(res.data.data, context);
        replyFunc(context, payload);
    }).catch(err => {
        console.error(err.response);
        replyFunc(context, "出错啦！");
    });
}

/**
 * 通过用户名添加订阅
 * @param {string} name 微博用户名
 * @param {string} option_nl 偏好设置，可以是"仅原创"，"包含转发"，"仅带图"
 * @param {object} context
 * @returns {boolean} 成功返回true
 */
async function addSubByName(name, option_nl, context) {
    let uid = await getUserId(name);
    if (!uid) {
        replyFunc(context, "没这人", true);
        return true;
    }
    else {
        let option = optnl2opt(option_nl)
        subscribe(uid, option, context);
        return false;
    }
}

/**
 * 通过用户uid添加订阅
 * @param {string} url 单条微博url m.weibo.cn
 * @param {string} option_nl 偏好设置，可以是"仅原创"，"包含转发"，"仅带图"
 * @param {object} context
 */
function addSubByUid(url, option_nl, context) {
    axios.get(url, {params : httpHeader().headers}).then(res => {
        let temp = /https:\/\/m.weibo.cn(\/(\d+)\/\d+|\/u\/(\d+))/.exec(url);
        let uid = temp[2] ? temp[2] : temp[3]
        let option = optnl2opt(option_nl);
        subscribe(uid, option, context);
    }).catch(err => {
        replyFunc(context, "无法订阅这个人", true);
        console.error(err);
    });
}

/**
 * @param {string} name 微博用户名
 * @param {number} num 需要获取的微博，-1为查找最新，-2为查找置顶，0为置顶或者最新，1是次新，以此类推，最高到9
 * @returns {} no return
 */
function rtWeibo(name, num, context) {
    getUserId(name).then(uid => {
        if (uid) getTimeline(uid, num).then(res => {
            format(res, context).then(payload => {
                replyFunc(context, payload);
            }).catch(err => {console.error(err); replyFunc(context, "中途错误", true);});
        }).catch(err => {console.error(err); replyFunc(context, "等下再试", true);});
        else replyFunc(context, "查无此人", true);
    }).catch(err => {console.error(err); replyFunc(context, "中途错误", true);});
}

/**
 * @param {object} context 
 * @returns {boolean} 如果是这里的功能，返回true，否则为false
 */
function weiboAggr(context) {
    if (!CONFIG.enable) return false;
    if (/^看看(.+?)的?((第[0-9]?[一二三四五六七八九]?条)|(上*条)|(置顶)|(最新))?微博/.test(context.message)) {	
		let num = 1;
        let name = "";
        if (/置顶/.test(context.message)) (num = -1);
        else if (/最新/.test(context.message)) (num = 0);
        else if (/上上上条/.test(context.message)) (num = 3);
        else if (/上上条/.test(context.message)) (num = 2);
        else if (/上条/.test(context.message)) (num = 1);
	    else if (/第.+?条/.test(context.message)) {
            let temp = /第([0-9]|[一二三四五六七八九])条/.exec(context.message)[1];
            if (temp==0 || temp=="零") (num = -1);
            else if (temp==1 || temp=="一") (num = 0);
            else if (temp==2 || temp=="二") (num = 1);
            else if (temp==3 || temp=="三") (num = 2);
            else if (temp==4 || temp=="四") (num = 3);
            else if (temp==5 || temp=="五") (num = 4);
            else if (temp==6 || temp=="六") (num = 5);
            else if (temp==7 || temp=="七") (num = 6);
            else if (temp==8 || temp=="八") (num = 7);
            else if (temp==9 || temp=="九") (num = 8);
        }
        else num = -1;       
        name = /看看(.+?)的?((第[0-9]?[一二三四五六七八九]?条)|(上*条)|(置顶)|(最新))?微博/.exec(context.message)[1];
        rtWeibo(name, num, context);
        return true;
	}
    else if (/^看看\s?https:\/\/(m\.weibo\.cn\/(detail|status|\d+)\/\d+$|(www\.)?weibo\.com\/\d+\/[A-Za-z0-9]{9}$)/.test(context.message)) {
        let id = /com\/\d+\/([A-Za-z0-9]{9})|cn\/\d+\/(\d+)|detail|status\/(\d+)/.exec(context.message)
            .filter((noEmpty) => {return noEmpty != undefined})[1];
        rtSingleWeibo(id, context);
        return true;
    }
    else if (/^\[CQ:json.+"appID":"100736903"/.test(context.message)) {
        let id = /https:\/\/m\.weibo\.cn\/status\/(\d+)/.exec(context.message)[1];
        rtSingleWeibo(id, context);
        return true;
    }
    else if (/^订阅.+的?微博([>＞](仅转发|只看图|全部))?/.test(context.message)) {
        if (!global.permissionCheck(context, PERMISSION)) return true;
        let {groups : {name, option_nl}} = /订阅(?<name>.+)的?微博([>＞](?<option_nl>仅转发|只看图|全部))?/.exec(context.message);
        if (option_nl == undefined) option_nl = "仅原创";
        addSubByName(name, option_nl, context);
        return true;
    }
    else if (/^订阅微博\s?https:\/\/m.weibo.cn.+([>＞](仅转发|只看图|全部))?/.test(context.message)) {
        if (!global.permissionCheck(context, PERMISSION)) return true;
        let {groups : {url, option_nl}} = /(?<url>https:\/\/m.weibo.cn.+)([>＞](?<option_nl>仅转发|只看图|全部))?/.exec(context.message);
        if (option_nl == undefined) option_nl = "仅原创";
        addSubByUid(url, option_nl, context);
        return true;
    }
    else if (/^取消订阅.+的?微博$/.test(context.message)) {
        if (!global.permissionCheck(context, PERMISSION)) return true;
        let name = /取消订阅(.+)的?微博/i.exec(context.message)[1];
        unSubscribe(name, context);
        return true;
    }
    else if (/^查看(订阅微博|微博订阅)$/.test(context.message)) {
        checkWeiboSubs(context);
        return true;
    }
    else if (/^清空微博订阅$/.test(context.message)) {
        if (global.permissionCheck(context, ["SU", "owner"])) clearSubs(context, context.group_id);
        return true;
    }
    else return false;
}

export default {weiboAggr, checkWeiboDynamic, clearSubs};